packages/common/src/failure.ts

```diff
@@ -446,122 +327,320 @@ export function ensureTemporalFailure(err: unknown): TemporalFailure {
 }
 
 /**
- * Converts a Failure proto message to a JS Error object if defined or returns undefined.
+ * Get the root cause message of given `error`.
+ *
+ * In case `error` is a {@link TemporalFailure}, recurse the `cause` chain and return the root `cause.message`.
+ * Otherwise, return `error.message`.
  */
-export function optionalFailureToOptionalError(
-  failure: ProtoFailure | undefined | null,
-  payloadConverter: PayloadConverter
-): TemporalFailure | undefined {
-  return failure ? failureToError(failure, payloadConverter) : undefined;
+export function rootCause(error: unknown): string | undefined {
+  if (error instanceof TemporalFailure) {
+    return error.cause ? rootCause(error.cause) : error.message;
+  }
+  if (error instanceof Error) {
+    return error.message;
+  }
+  if (typeof error === 'string') {
+    return error;
+  }
+  return undefined;
 }
 
 /**
- * Converts a Failure proto message to a JS Error object.
+ * A `FailureConverter` is responsible to convert from proto `Failure` instances to JS `Errors` and back.
  *
- * Does not set common properties, that is done in {@link failureToError}.
+ * It is recommended to use the {@link DefaultFailureConverter} and not attempt to customize the default implementation
+ * in order to maintain cross language failure serialization compatibility.
```

Comment:
This comment should explain why this is exposed at all then, and when you might need to customize it

Reply:
There shouldn't ever be a need for that

Reply:
Then why is this exposed at all? If that's the case I find this comment even more confusing

Reply:
To give the full flexibility, we could have just exposed a boolean to decide whether or not to encode failure attributes but that would limit power users that might want to implement their own failure serialization logic